storage/storage.go

```diff
@@ -46,8 +53,78 @@ type Handler interface {
 	RemoveSortedSetRange(string, string, string) error
 }
 
+const defaultHashAlgorithm = "murmur64"
+
+// If hashing algorithm is empty, use legacy key generation
+func GenerateToken(orgID, keyID, hashAlgorithm string) (string, error) {
+	if keyID == "" {
+		keyID = strings.Replace(uuid.NewV4().String(), "-", "", -1)
+	}
+
+	if hashAlgorithm != "" {
+		_, err := hashFunction(hashAlgorithm)
+		if err != nil {
+			hashAlgorithm = defaultHashAlgorithm
+		}
+
+		jsonToken := fmt.Sprintf(`{"org":"%s","id":"%s","h":"%s"}`, orgID, keyID, hashAlgorithm)
+		return base64.StdEncoding.EncodeToString([]byte(jsonToken)), err
+	}
+
+	// Legacy keys
+	return orgID + keyID, nil
+}
+
+// `{"` in base64
+const B64JSONPrefix = "ey"
+
+func TokenHashAlgo(token string) string {
+	// Legacy tokens not b64 and not JSON records
+	if strings.HasPrefix(token, B64JSONPrefix) {
+		if jsonToken, err := base64.StdEncoding.DecodeString(token); err == nil {
+			hashAlgo, _ := jsonparser.GetString(jsonToken, "h")
+			return hashAlgo
+		}
+	}
+
+	return ""
+}
+
+func TokenOrg(token string) string {
+	if strings.HasPrefix(token, B64JSONPrefix) {
+		if jsonToken, err := base64.StdEncoding.DecodeString(token); err == nil {
+			// Checking error in case if it is a legacy tooken which just by accided has the same b64JSON prefix
+			if org, err := jsonparser.GetString(jsonToken, "org"); err == nil {
+				return org
+			}
+		}
+	}
+
+	// 24 is mongo bson id length
+	if len(token) > 24 {
+		return token[:24]
+	}
+
+	return ""
+}
+
+func hashFunction(algorithm string) (hash.Hash, error) {
+	switch algorithm {
+	case "sha256":
+		return sha256.New(), nil
+	case "murmur64":
+		return murmur3.New64(), nil
+	case "murmur128":
+		return murmur3.New128(), nil
+	case "", "murmur32":
+		return murmur3.New32(), nil
+	default:
+		return murmur3.New32(), fmt.Errorf("Unknown key hash function: %s. Falling back to murmur32.", algorithm)
+	}
+}
+
 func HashStr(in string) string {
-	h := murmur3.New32()
+	h, _ := hashFunction(TokenHashAlgo(in))
```

Comment:
method `HashStr ` is heavily used when key hashing enabled, correct me if I am wrong - we always do `base64.StdEncoding.EncodeToString` when we generateToken from key, org ID and then have to do `base64.StdEncoding.DecodeString` right away to get from storage or update from storage. this base64 encoding/decoding seems to be extra work to me